Cython/Utility/Optimize.c

```diff
@@ -1325,17 +1325,31 @@ static {{c_ret_type}} __Pyx_Unpacked_{{cfunc_name}}(PyObject *op1, PyObject *op2
                 x = q;
             }
         {{else}}
+            {{if op == 'Rshift' or op == 'Lshift'}}
+#if !(defined(__cplusplus) && __cplusplus < 202002L)
+            // This is well-defined only in C++20
+            if (unlikely(a < 0)) goto fallback;
+#endif
+            {{endif}}
             x = a {{c_op}} b;
-            {{if op == 'Lshift'}}
+            {{if op == 'Lshift' or op == 'Rshift'}}
 #ifdef HAVE_LONG_LONG
-            if (unlikely(!(b < (long) (sizeof(long)*8) && a == x >> b)) && a) {
+            if (unlikely(b >= (long) (sizeof(long)*8))
```

Comment:
This needs changing - rshift too far has a known answer so I may as well return that rather than falling back to Python